samples/gen-go/client/doer.go

```diff
@@ -58,45 +59,83 @@ func (d tracingDoer) Do(c *http.Client, r *http.Request) (*http.Response, error)
 
 // retryHandler retries 50X http requests
 type retryDoer struct {
-	d              doer
-	defaultRetries int
+	d           doer
+	retryPolicy RetryPolicy
 }
 
-// WithRetries returns a new context that overrides the number of retries to do for a particular
-// request.
-func WithRetries(ctx context.Context, retries int) context.Context {
-	return context.WithValue(ctx, retryContext{}, retries)
+// RetryPolicy defines a retry policy.
+type RetryPolicy interface {
+	// Backoffs returns the number and timing of retry attempts.
+	Backoffs() []time.Duration
+	// Retry receives the http request, as well as the result of
+	// net/http.Client's `Do` method.
+	Retry(*http.Request, *http.Response, error) bool
 }
 
-// retryContext is the key the retry configuration. For demonstration purposes it's just a count
-// of the number of retries right now.
-type retryContext struct{}
-
-func (d *retryDoer) Do(c *http.Client, r *http.Request) (*http.Response, error) {
+// DefaultRetryPolicy defines the default retry policy.
+type DefaultRetryPolicy struct{}
 
-	resp, err := d.d.Do(c, r)
-	if err != nil {
-		return resp, err
+// Backoffs returns five backoffs with exponentially increasing wait times
+// between requests: 100, 200, 400, 800, and 1600 milliseconds +/- up to 5% jitter.
+func (DefaultRetryPolicy) Backoffs() []time.Duration {
+	ret := make([]time.Duration, 5)
+	next := 100 * time.Millisecond
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+	e := 0.05 // +/- 5 percent jitter
+	for i := range ret {
+		ret[i] = next + time.Duration(((rnd.Float64()*2)-1)*e*float64(next))
+		next *= 2
 	}
+	return ret
+}
 
-	// If the request can't be retried then just return immediately. We retry all idempotent
-	// http requests. The only two that can't be retried are post and patch
-	if r.Method == "POST" || r.Method == "PATCH" {
-		return resp, err
+// Retry will retry non-POST, non-PATCH requests that 5XX.
+// TODO: It does not currently retry any errors returned by net/http.Client's `Do`.
```

Comment:
fyi, here's the Go transport retrying I mentioned https://github.com/golang/go/blob/master/src/net/http/transport.go#L403-L450